agent/containermetadata/parse_metadata.go

```diff
@@ -0,0 +1,136 @@
+			networkInterfaceList = append(networkInterfaceList, networkInterface)
+		}
+	} else {
+		networkInterface := Network{NetworkMode: networkModeFromHostConfig, IPv4Address: ipv4AddressFromSettings}
+		networkInterfaceList = append(networkInterfaceList, networkInterface)
+	}
+
+	return NetworkMetadata{
+		networks: networkInterfaceList,
+	}, nil
+}
+
+// parseDockerContainerMetadata parses the metadata in a docker container
+// and packages this data for JSON marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this stage.
+func parseDockerContainerMetadata(taskARN string, containerName string, dockerContainer *docker.Container) DockerContainerMetadata {
+	if dockerContainer == nil {
+		seelog.Warnf("Failed to parse container metadata for task %s container %s: container metadata not available or does not exist", taskARN, containerName)
+		return DockerContainerMetadata{}
+	}
+
+	// In most cases a container should never lack a config but we check regardless to avoid
+	// nil pointer exceptions (Could occur if there is some error in the docker api call, if the
+	// container we receive has incomplete information)
+	imageNameFromConfig := ""
+	if dockerContainer.Config != nil {
+		imageNameFromConfig = dockerContainer.Config.Image
+	} else {
+		seelog.Warnf("Failed to parse container metadata for task %s container %s: container has no configuration", taskARN, containerName)
+	}
+
+	// Get Port bindings from docker configurations
+	var ports []api.PortBinding
+	var err error
+	if dockerContainer.HostConfig != nil {
+		ports, err = api.PortBindingFromDockerPortBinding(dockerContainer.HostConfig.PortBindings)
+		if err != nil {
+			seelog.Warnf("Failed to parse container metadata for task %s container %s: %v", taskARN, containerName, err)
+		}
+	} else {
+		seelog.Warnf("Failed to parse container metadata for task %s container %s: container has no host configuration", taskARN, containerName)
+	}
+
+	networkMetadata, err := parseNetworkMetadata(dockerContainer.NetworkSettings, dockerContainer.HostConfig)
+	if err != nil {
+		seelog.Warnf("Failed to parse container metadata for task %s container %s: %v", taskARN, containerName, err)
+	}
+
+	return DockerContainerMetadata{
+		containerID:         dockerContainer.ID,
+		dockerContainerName: dockerContainer.Name,
+		imageID:             dockerContainer.Image,
+		imageName:           imageNameFromConfig,
+		ports:               ports,
+		networkInfo:         networkMetadata,
+	}
+}
+
+// parseMetadata gathers metadata from a docker container, and task
+// configuration and data then packages it for JSON Marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this or the above stage.
+func (manager *metadataManager) parseMetadata(dockerContainer *docker.Container, taskARN string, containerName string) Metadata {
```

Comment:
Generally, I'd like to see the functions in a file organized with caller-above-callee so that you start with abstractions at a high level and get deeper into details as you read further in the file.  That would mean that the entrypoints (`parseMetadata` and `parseMetadataAtContainerCreate`) are at the top, then `parseDockerContainerMetadata`, and last `parseNetworkMetadata`.